Controllers

zagliadaAA opened the pull request.

Commit:
[controllers] добавлен первый контроллер

Commit:
[controllers] создан контроллер create

Commit:
[controllers] исправлено название проекта

Commit:
[controllers] добавлены контроллеры для client_usecase и medical_report_usecase

popkovvvv commented on cmd/service_provider/routes.go:
```diff
@@ -0,0 +1,20 @@
+	// медицинское заключение
+	mux.HandleFunc("POST /reports", sp.getMedicalReportController().Create)
+	mux.HandleFunc("DELETE /reports/{id}", sp.getMedicalReportController().Delete)
+	mux.HandleFunc("PUT /reports", sp.getMedicalReportController().Update)
```
а читать никто заключение не захочет?
и список клиентов?

popkovvvv commented on internal/controller/client_controller/create.go:
```diff
@@ -0,0 +1,92 @@
+		PhoneNumber: req.PhoneNumber,
+	})
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
```
нужно сделать объект ошибки для StatusBadRequest

popkovvvv commented on internal/controller/client_controller/create.go:
```diff
@@ -0,0 +1,92 @@
+
+	encoder := json.NewEncoder(w)
+	w.Header().Set("Content-Type", "application/json; charset=utf-8")
+	err = encoder.Encode(createClientResp{
```
почему бы нам не сделать что то подобное для респоса?

if err := c.DecodeRequest(w, r, &req); err != nil {
		return
	}

popkovvvv commented on internal/controller/client_controller/delete.go:
```diff
@@ -0,0 +1,29 @@
+	}
+
+	w.WriteHeader(http.StatusOK)
+	_, err = w.Write([]byte("Client deleted successfully"))
```
для delete тело не нужно возвращать, только 200 статус

popkovvvv commented on internal/controller/client_controller/delete.go:
```diff
@@ -0,0 +1,29 @@
+	w.WriteHeader(http.StatusOK)
+	_, err = w.Write([]byte("Client deleted successfully"))
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
```
должен быть объект а не строка

popkovvvv commented on internal/controller/client_controller/update.go:
```diff
@@ -0,0 +1,96 @@
+	idString := r.URL.Path[len("/clients/"):]
+	id, err := strconv.Atoi(idString)
+	if err != nil {
+		http.Error(w, "invalid ID client", http.StatusBadRequest)
```
вроде как для валидации у нас есть ошибка нужная

popkovvvv commented on internal/controller/medical_report_controller/controller.go:
```diff
@@ -0,0 +1,36 @@
+	}
+}
+
+func (c *Controller) DecodeRequest(w http.ResponseWriter, r *http.Request, req interface{}) error {
```
перенести в internal/controller/http_utils.go

popkovvvv commented on internal/controller/medical_report_controller/update.go:
```diff
@@ -0,0 +1,80 @@
+
+	encoder := json.NewEncoder(w)
+	w.Header().Set("Content-Type", "application/json; charset=utf-8")
+	err = encoder.Encode(updateMedicalReportResp{
```
давай в папке internal/controller/medical_report_controller  сделаем файлик mapper где будем кастить доменную сущность в респонсы и туту просто будем вызывать 

ну и касается второго контроллера

Commit:
[controllers] перенесены DecodeRequest и EncodeResponse, добавлен объект ошибки для StatusBadRequest

Commit:
[controllers] добавлен объект ошибки для StatusBadRequest,
изменены методы getReportByIdClient и findById,
исправлены тесты,
добавлены usecase для методов GetClientByID и GetReportById,
добавлены тесты для GetClientByID и GetReportById,
добавлены mapper для контроллеров.

popkovvvv commented on internal/controller/client_controller/delete.go:
```diff
@@ -0,0 +1,27 @@
+	idString := r.URL.Path[len("/clients/"):]
+	id, err := strconv.Atoi(idString)
+	if err != nil {
+		controller.RespondStatusBadRequestError(w, controller.NewStatusBadRequestError("failed converted to type int"))
```
по идее это валидация

popkovvvv commented on internal/controller/client_controller/delete.go:
```diff
@@ -0,0 +1,27 @@
+)
+
+func (c *Controller) Delete(w http.ResponseWriter, r *http.Request) {
+	idString := r.URL.Path[len("/clients/"):]
```
так не достаем

// Извлечение ID из пути
```
	id := r.PathValue("id")
	clientID, err := strconv.ParseInt(id, 0, 64)
	if err != nil {
		controller.ValidationErrorRespond(w, controller.NewValidationError("client id not present", "id"))

		return
	}
```

Commit:
[controllers] изменено получение id из get запроса
